boa_engine/src/vm/trace.rs

```diff
@@ -0,0 +1,289 @@
+/// for all possible compiled output
+pub struct VmTrace {
+    options: Cell<TraceOptions>,
+    tracer: Box<dyn Tracer>,
+}
+
+// ==== Public API ====
+
+impl VmTrace {
+    #[must_use]
+    /// Create a partial `VmTrace`.
+    pub fn partial() -> Self {
+        Self {
+            options: Cell::new(TraceOptions::empty()),
+            tracer: Box::new(DefaultTracer),
+        }
+    }
+
+    #[must_use]
+    /// Method for adding a compiled action on initialization.
+    pub fn with_tracer(mut self, tracer: Box<dyn Tracer>) -> Self {
+        self.set_tracer(tracer);
+        self
+    }
+
+    /// Sets the current `Tracer` of `VmTrace`.
+    pub fn set_tracer(&mut self, tracer: Box<dyn Tracer>) {
+        self.tracer = tracer;
+    }
+}
+
+// ==== Internal VmTrace methods ====
+
+impl VmTrace {
+    /// Returns if Trace type is a complete trace.
+    pub(crate) fn is_full_trace(&self) -> bool {
+        self.options.get().contains(TraceOptions::FULL_TRACE)
+    }
+
+    /// Returns if the trace is only a partial one.
+    pub fn is_partial_trace(&self) -> bool {
+        !self.is_full_trace()
+    }
+
+    /// Returns if the a partial trace has been determined to be active.
+    pub fn is_active(&self) -> bool {
+        self.options.get().contains(TraceOptions::ACTIVE)
+    }
+
+    /// Sets the `ACTIVE` bitflag to true.
+    pub(crate) fn activate(&self) {
+        let mut flags = self.options.get();
+        flags.set(TraceOptions::ACTIVE, true);
+        self.options.set(flags);
+    }
+
+    /// Sets the `ACTIVE` flag to false.
+    pub(crate) fn inactivate(&self) {
+        let mut flags = self.options.get();
+        flags.set(TraceOptions::ACTIVE, false);
+        self.options.set(flags);
+    }
+
+    /// Returns whether a trace should run on an instruction.
+    pub(crate) fn should_trace(&self) -> bool {
+        self.is_full_trace() || self.is_active()
+    }
+}
+
+// ==== Trace Event/Action Methods ====
+
+impl VmTrace {
+    const COLUMN_WIDTH: usize = 26;
+    const TIME_COLUMN_WIDTH: usize = Self::COLUMN_WIDTH / 2;
+    const OPCODE_COLUMN_WIDTH: usize = Self::COLUMN_WIDTH;
+    const OPERAND_COLUMN_WIDTH: usize = Self::COLUMN_WIDTH;
+    const NUMBER_OF_COLUMNS: usize = 4;
+
+    /// Trace the current `CallFrame` according to current state
+    pub(crate) fn trace_call_frame(&self, vm: &Vm, interner: &Interner) {
+        if self.is_full_trace() {
+            self.trace_compiled_bytecode(vm, interner);
+            self.call_frame_header(vm);
+        } else if self.is_partial_trace() && vm.frame().code_block().traceable() {
+            if !vm.frame().code_block().frame_traced() {
+                self.trace_current_bytecode(vm, interner);
+                vm.frame().code_block().set_frame_traced(true);
+            }
+            self.call_frame_header(vm);
+            self.activate();
+        } else {
+            self.call_frame_header(vm);
+        }
```

Comment:
Still need to work on this portion. The rest is rebased, so it shouldn't take too long 😄

Reply:
Added an enum (went with `TraceAction` over `ShouldTrace`), but now some of that logic is loaded into `should_trace`. Let me know what you think.

Reply:
Really nice! Though, I think I didn't explain myself good enough. The goal of adding that enum was so that we can move the "should be traced partially or fully" logic from `VmTracer` to the tracers itself, by adding a `Tracer::should_trace(&self, code: Gc<CodeBlock>) -> TraceAction` method. Then, implementors would be the ones taking care of choosing when to just print the instructions vs the compiled bytecode.

Reply:
Should we be exposing that much of the `Vm` internals to `Tracer`? Part of the current implementation is to only expose preformatted strings in a bit more of a "controlled environment" so to speak, which included handling the logic aspect.

Reply:
I'm pretty sure we already expose `CodeBlock` publicly, but users cannot do much with it aside from getting its address and name.

Reply:
I'm playing around with the update, and it's nicer in some places, but feels a lot more convoluted in others. I'm also not entirely sure on how to calling multiple `$boa.function.traceable()` might work as calling the function would remove the previous `Tracer`, which may cause problems on generators and async functions.

Reply:
I'd say that we should try to find an architecture that allows us to always use the same `Tracer`, mostly for simplicity.